### PR TITLE
Implemented: login and logout functionality and used pinia instead for vuex store #4

### PR DESCRIPTION
**Pull request by Ritika-Patel08**
### Related Issues

Related Issue #4 

### Short Description and Why It's Useful

- Implemented login and logout functionality.
- Implemented store using pinia for state management 
- Removed the vuex store and its related files.
- Removed dxp components and related fucntions 

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/hotwax/napita-flow-manager#contribution-guideline)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,11 +21,13 @@
     "@ionic/vue-router": "~7.6.0",
     "@types/file-saver": "^2.0.4",
     "@types/papaparse": "^5.3.1",
+    "axios": "^0.21.4",
     "boon-js": "^2.0.3",
     "core-js": "^3.6.5",
     "file-saver": "^2.0.5",
     "luxon": "^2.3.0",
     "mitt": "^2.1.0",
+    "pinia": "^2.1.7",
```
Comment (ymaheshwari1) [inline package.json:30]: Remove vuex and vuex-persistedstate

File: src/api/index.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
+import axios from "axios";
+import { setupCache } from "axios-cache-adapter"
+import OfflineHelper from "@/offline-helper"
+import emitter from "@/event-bus"
+import { StatusCodes } from "http-status-codes";
+import router from "@/router"
+import { useUserStore } from '@/store/user';
+
+// axios.interceptors.request.use((config: any) => {
```
Comment (ymaheshwari1) [inline src/api/index.ts]: We need to pass the token when making api calls using `api` method

File: src/api/index.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  if(customConfig.cache) config.adapter = axiosCache.adapter;
+  const networkStatus =  await OfflineHelper.getNetworkStatus();
+  if (customConfig.queue && !networkStatus.connected) {
+    if (!config.headers) config.headers = { ...axios.defaults.headers.common, ...config.headers };
+    emitter.emit("queueTask", {
+      callbackEvent: customConfig.callbackEvent,
+      payload: config
+    });
+  } else {
+      return axios(config);
```
Comment (ymaheshwari1) [inline src/api/index.ts]: Improve indentation

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/api/index.ts (hunk 3 above)
Comment (Ritika-Patel08) [reply src/api/index.ts]: done.

File: package.json (hunk 1 above)
Comment (Ritika-Patel08) [reply package.json:30]: Done.

File: src/App.vue [hunk 4]
```diff
@@ -1,124 +1,55 @@
...
-      userProfile: 'user/getUserProfile'
-    })
-  },
-  methods: {
-    async presentLoader(options = { message: '', backdropDismiss: true }) {
-      // When having a custom message remove already existing loader
-      if(options.message && this.loader) this.dismissLoader();
+const loader = ref(null) as any
+//const userProfile = computed(() => store.getters["user/getUserProfile"])
+const router = useRouter();
```
Comment (amansinghbais) [inline src/App.vue]: Remove unused router declaration.

File: src/main.ts [hunk 5]
```diff
@@ -3,6 +3,8 @@ import App from './App.vue'
 import router from './router';
 import logger from './logger';
 import { showToast } from '@/utils'
```
Comment (amansinghbais) [inline src/main.ts]: Remove showToast import if not used.

File: src/router/index.ts [hunk 6]
```diff
@@ -43,32 +37,25 @@ const routes: Array<RouteRecordRaw> = [
   {
     path: '/login',
     name: 'Login',
-    component: DxpLogin,
+    component: Login,
     beforeEnter: loginGuard
+
```
Comment (amansinghbais) [inline src/router/index.ts]: Remove this unwated space.

File: src/router/index.ts [hunk 7]
```diff
@@ -43,32 +37,25 @@ const routes: Array<RouteRecordRaw> = [
...
   },
   {
     path: "/settings",
     name: "Settings",
     component: Settings,
     beforeEnter: authGuard
-  }
+  },
```
Comment (amansinghbais) [inline src/router/index.ts]: Remove unwanted change

File: src/router/index.ts [hunk 8]
```diff
@@ -43,32 +37,25 @@ const routes: Array<RouteRecordRaw> = [
...
-    // If the user has navigated from Login page or if it is page load, redirect user to settings page without showing any toast
-    if (redirectToPath == "/login" || redirectToPath == "/") redirectToPath = "/settings";
-    else showToast(translate('You do not have permission to access this page'));
-    return {
-      path: redirectToPath,
-    }
-  }
-})
+// router.beforeEach((to, from, next) => {
+//   authGuard(to, from, next);
```
Comment (amansinghbais) [inline src/router/index.ts]: Remove this if not needed.

File: src/services/UserService.ts [hunk 9]
```diff
@@ -1,216 +1,61 @@
...
+import { useUserStore } from '@/store/user';
+import api, { client } from "@/api"
+
+const login = async (username: string, password: string): Promise<any> => {
+  const userStore = useUserStore();
+  const baseURL = userStore.getBaseUrl;
+  const params = new URLSearchParams();
+  params.append('username', username);
+  params.append('password', password);
+  //let token = ""
```
Comment (amansinghbais) [inline src/services/UserService.ts]: Remove token variable if not needed

File: src/api/index.ts (hunk 3 above)
Comment (amansinghbais) [reply src/api/index.ts]: Not fixed re-check

File: src/components/Tabs.vue [hunk 10]
```diff
@@ -0,0 +1,48 @@
+<template>
+  <ion-page>
+    <ion-tabs>
+      <ion-router-outlet></ion-router-outlet>
+      <ion-tab-bar slot="bottom" v-if="showFooter()">
+
```
Comment (amansinghbais) [inline src/components/Tabs.vue]: Remove unnecessary space.

File: src/store/user.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
+import { defineStore } from "pinia";
+import { UserService } from "@/services/UserService"
+import { DateTime } from 'luxon';
+
+export const useUserStore = defineStore('user', {
+  state: () => {
+    return {
+      instanceUrl: "",
```
Comment (amansinghbais) [inline src/store/user.ts]: Follow consistent structure for empty string declaration.

File: src/views/Settings.vue [hunk 12]
```diff
@@ -11,268 +11,63 @@
...
               <Image :src="userProfile.partyImageUrl"/>
-            </ion-avatar>
+            </ion-avatar> -->
             <!-- ion-no-padding to remove extra side/horizontal padding as additional padding 
             is added on sides from ion-item and ion-padding-vertical to compensate the removed
             vertical padding -->
             <ion-card-header class="ion-no-padding ion-padding-vertical">
-              <ion-card-subtitle>{{ userProfile?.userLoginId }}</ion-card-subtitle>
-              <ion-card-title>{{ userProfile?.partyName }}</ion-card-title>
+              <ion-card-subtitle>{{ translate("demo-user") }}</ion-card-subtitle>
```
Comment (amansinghbais) [inline src/views/Settings.vue]: No need to translate this.

File: src/views/Settings.vue [hunk 13]
```diff
@@ -11,268 +11,63 @@
       <div class="user-profile">
         <ion-card>
           <ion-item lines="full">
-            <ion-avatar slot="start" v-if="userProfile?.partyImageUrl">
+            <!-- <ion-avatar slot="start" v-if="userProfile?.partyImageUrl">
```
Comment (amansinghbais) [inline src/views/Settings.vue]: Remove following code if not required.

File: src/store/user.ts [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    },
+    async logout() { 
+      try {
+        await UserService.logout();
+        this.token.value = '';
+        this.token.expirationTime = '';
+        this.instanceUrl = '';
+        return Promise.resolve(); 
+      } catch (err) {
+        console.error('Error logging out:', err);
```
Comment (amansinghbais) [inline src/store/user.ts]: Check whether logger is implemented in app or not? If implemented, use that.

**Review by amansinghbais — APPROVED** [review]

File: src/services/UserService.ts [hunk 15]
```diff
@@ -1,216 +1,60 @@
...
-      }
-    });
-    if(hasError(resp)) return Promise.reject("Error getting user profile: " + JSON.stringify(resp.data));
-    return Promise.resolve(resp.data)
-  } catch(error: any) {
-    return Promise.reject(error)
+    }) as any;  
+    return Promise.resolve(resp);
+  } catch (err) {
+    return Promise.reject("Sorry, login failed. Please try again");
```
Comment (ymaheshwari1) [inline src/services/UserService.ts]: I think we should allow user to login into the app without having expiration time check, simply login the user into the app without expirationTime.

File: src/views/Login.vue [hunk 16]
```diff
@@ -0,0 +1,77 @@
+<template>
+    <ion-page>
```
Comment (ymaheshwari1) [inline src/views/Login.vue]: Indentaion Issue

File: src/views/Settings.vue [hunk 17]
```diff
@@ -11,268 +11,57 @@
...
-            <ion-avatar slot="start" v-if="userProfile?.partyImageUrl">
-              <Image :src="userProfile.partyImageUrl"/>
-            </ion-avatar>
-            <!-- ion-no-padding to remove extra side/horizontal padding as additional padding 
-            is added on sides from ion-item and ion-padding-vertical to compensate the removed
-            vertical padding -->
             <ion-card-header class="ion-no-padding ion-padding-vertical">
-              <ion-card-subtitle>{{ userProfile?.userLoginId }}</ion-card-subtitle>
-              <ion-card-title>{{ userProfile?.partyName }}</ion-card-title>
+              <ion-card-subtitle>demo-user</ion-card-subtitle>
```
Comment (ymaheshwari1) [inline src/views/Settings.vue]: Make this dynamic.

File: src/store/user.ts [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+      if (state.token.expirationTime ) {
+        const currTime = DateTime.now().toMillis()
+        isTokenExpired = state.token.expirationTime < currTime
+      }
+      return state.token.value && !isTokenExpired
+    }
+  },
+  actions: {
+    async login( username: string, password:string ) {
+      try {
```
Comment (ymaheshwari1) [inline src/store/user.ts:47]: Add loader to let user know that something is going on in the background for all actions wherever required.

File: src/api/index.ts [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+  // Cannot store it in cookies or localStorage as its not safe
+  // https://stackoverflow.com/questions/67062876/is-it-secure-to-store-a-csrf-token-value-in-the-dom
+  // https://stackoverflow.com/questions/62289684/what-is-the-correct-way-for-a-client-to-store-a-csrf-token
+  const csrfToken = response.headers["x-csrf-token"]
+  const meta = document.createElement("meta")
+  meta.name = "csrf"
+  meta.content = csrfToken
+  document.getElementsByTagName("head")[0].appendChild(meta)
+
+  document.cookie = `x-csrf-token=${csrfToken}`
```
Comment (ymaheshwari1) [inline src/api/index.ts]: Remove this code related to csrf token.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/views/Login.vue [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+function login() {
+  userStore.setUserInstanceUrl(instanceUrl.value.trim());
+  userStore.login( username.value.trim(), password.value ).then((data: any) => {
+    if (data) {
+      username.value = ""
+      password.value = ""
+      router.push("/")
+    }
+  }).catch(err => err)
```
Comment (ymaheshwari1) [inline src/views/Login.vue:59]: Improve indentation

File: src/store/user.ts [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+        const currTime = DateTime.now().toMillis()
+        isTokenExpired = state.token.expirationTime < currTime
+      }
+      return state.token.value && !isTokenExpired
+    }
+  },
+  actions: {
+    async login(username: string, password: string) {
+    
+      emitter.emit('presentLoader', { message: 'Logging in', backdropDismiss: false })
```
Comment (ymaheshwari1) [inline src/store/user.ts]: Move this loader to the calling method

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/store/user.ts [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    async setUserInstanceUrl(instanceUrl: string) {
+      this.instanceUrl = instanceUrl;
+    },
+    async logout() { 
+      try {
+        await UserService.logout();
+        this.token.value = '';
+        this.token.expirationTime = '';
+        this.instanceUrl = '';
+        emitter.emit('dismissLoader')
```
Comment (ymaheshwari1) [inline src/store/user.ts:77]: We have not emitted the presentLoader function before dismissing.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

**Review by ymaheshwari1 — APPROVED** [review]

**Pull request merged**